dma: cAVS HDA DMA Driver

teburd opened the pull request:
Adds HDA host in and host out DMA drivers for Intel cAVS.

Each HDA stream direction is a driver here as the behavior needs to slightly vary. Most of the code is common code. Adding the link side drivers should be very easy.

A low level driver is available in the soc directory as a header, cavs_hda.h. This is meant to provide useful and direct access to manipulate registers. It enables unit testing with host side interaction without needing to map the hardware to the DMA API.

A DMA driver is built on top of the low level cavs_hda.h for host in and host out streams. In order to support the underlying ringbuffer the read_position and write_position were added to dma_status.

To inform the hardware of read or write changes to the underlying buffer, dma_reload may be called with the length of
the read/written bytes.

A quick mapping of how the DMA calls map to the existing SOF API would be...

From SOF to Zephyr...
 
dma_copy  ~= dma_reload
dma_status ~= dma_get_status

They aren't exactly the same, but very close.

Commit:
dma/cavs_hda: DMA driver for HDA on cAVS

Adds an initial driver for HDA streams on cAVS. A common code base is
provided for all HDA streams while the drivers are identified
differently as they have small behavior differences.

Uses dma_status to describe the positions for read/write. Uses dma_reload
to inform when to move the read/write positions. This closely follows
how HDA is being used in SoF

Simple test case is provided for both drivers.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

andyross commented on soc/xtensa/intel_adsp/common/include/cavs-link.ld:
```diff
@@ -410,12 +410,13 @@ SECTIONS {
-  .lpbuf (NOLOAD): {
+  /* dma buffers, require 128 byte alignment */
+  .hpbuf (NOLOAD): {
+    . = ALIGN(128);
```
Note that the modern SOF model is to allocate DMA memory from a unified heap, and the intent is that this will be extended via the multiheap model.  Putting buffers into static linker-managed sections is probably not what we want long term.  If you need to define a static buffer, just an "aligned(128)" attribute (or a soc-level macro to abstract that) is probably preferable.

I can't remember why the lpbuf section was left here.  IIRC there were some references to (but not uses of?) the section in the SOF source still, or something like that.

teburd replied:
It was used in the s1000 I believe with the designware dma, could be wrong here though. It seems nice to have a dedicated linker section to dma buffers to me and it was already there when I got here, I just tweaked it to move where it was placed.

Certainly another, more integrated, test case might do a heap allocation, send over a loop with hda, and validate it end to end with the DMA API involved, we'll get there.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+ */
+
+#ifndef HDA_HOST_OUT_BASE
+#define HDA_HOST_OUT_BASE 0x72800
```
This belong in devicetree.  Also the ifndef pattern here seems a little suspicious.  Why would you be worried about multiple definitions?  Sounds like maybe a header is missing an include guard?

teburd replied:
The intent here is yes, I'd move it to DTS when I go to implement the dma driver, its here because I needed something when testing. I can move this to my tests.

teburd replied:
I've removed the ifdefs, to remove the defines would mean I should remove the low level driver in the header I think, which I found incredibly useful.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+#define DGCS_BSCIE BIT(3) /* Buffer Segment Completion Interrupt Enable */
+
+/* Gateway Buffer Base Address */
+#define DGBBA(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x04))
```
Style thing: the rest of cavs likes to use C structs to define register blocks (a pattern I personally love, but will admit I didn't introduce).  Code in the rest of Zephyr tends to use offsets like this but with the sys_read/write() APIs.  Adding another scheme here ("macros returning pointer") is maybe a little confusing?

teburd replied:
This isn't the only driver to use a macro for a register returning a volatile pointer to the register. To be honest mapping this to structs, in my opinion, won't help build anymore clarity into the code here.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+ * @param sid Stream ID
+ */
+#define cavs_hda_dbg(name, base, sid)					\
+		printk("%s:%u %s(%u:0x%p), dgcs: 0x%x, dgbba 0x%x, "	\
```
Should probably be LOG_DBG().  Standard guidelines these days seem to be that printk() usage shouldn't be commited to new code at the driver level or above, it's for arch code, tests (some platforms don't have logging available) and ad-hoc debugging.  I actually like this rule because it makes for an easy way to grep to see that I haven't left any stray debug stuff in a PR.

Also: it's a macro, it should be CAVS_HDA_DBG() or CAVS_HDA_REG_DUMP() or something.  I think the all caps rule is actually a MISRA thing?

teburd replied:
Our style guide says "like linux" and linux has lower case macro functions all over the place. Kind of ugly to all cap this, but if its a MISRA thing I suppose its one of those things.

teburd replied:
I looked through the MISRA docs, and our docs, and I didn't see anything that specifically says macros must be capped.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+static inline int cavs_hda_set_buffer(uint32_t base, uint32_t sid,
+				     uint8_t *buf, uint32_t buf_size)
+{
+	void *cached_buf = arch_xtensa_cached_ptr(buf);
```
I'm confused here: why does the address need to be in the cached mapping?  Does the hardware require this?  Per spec, the DMA layer should be upstream of the xtensa mapping and shouldn't care, no?   If it doesn't work to use the uncached address, we should have a comment here calling that out.

teburd replied:
Its done so I can check the address is in L2  space which the hardware asks for. My interpretation may be wrong of the hardware docs but it does say cached or local L2. Local L2, I believe, is a different thing. Would the address range specified by cavs-mem L2_SRAM_BASE and L2_SRAM_SIZE be valid still if I chose uncached? That's my intent here, to ensure the address is in the range needed by the hardware to work.

teburd replied:
It's a 24 bit field which is fine, the goal in using the cached address is to validate the address is within the specified range and aligned appropriately. Fixed

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+
+/**
+ * @brief Determine the number of unused bytes in the fifo
+ *
```
Pedantry, but you should clarify that this is only useful for an **input**[1] stream ID.  It returns the wrong sense of subtraction if you're trying to figure out how long until an stream inbound to the DSP fills up.


[1] HDA terminology.  "Input" from the host side, output from the DSP, of course.

teburd replied:
Definitely, in the host output side it'd be useful to see the number of used bytes instead. SOF combines the two into I think something like hda_usable_bytes() which I could also do here.

andyross commented on soc/xtensa/intel_adsp/common/include/cavs_test.h:
```diff
@@ -0,0 +1,62 @@
+/* Copyright (c) 2022 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Nitpick: the motion of this header wants to be a separate patch.  Also this is really heavily coupled to the cavstool.py unit test script, so maybe a "cavstool.h" name or something might make it clearer what belongs here.  It's not just generic stuff for "testing".

teburd replied:
Done

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -18,11 +18,160 @@
 # Log is in the fourth window, they appear in 128k regions starting at 512k
 WINSTREAM_OFFSET = (512 + (3 * 128)) * 1024
 
+class HDAStream:
```
This looks like a perfect generalization of the stream from the limited output stream definition we had earlier.  But shouldn't we also be removing the older code from the loader and using this instead?

Also: this much surgery wants to be its own patch.  I'd suggest a refactoring that introduces this class and uses it instead of the older code, then dropping new protocols on top of it.

teburd replied:
To be frank, I was frightened by the time it might take to do just that because of the seemingly endless hardware quirks. But it definitely tries to abstract it, and yes I should refactor the existing firmware load into this to some degree.

teburd replied:
I'm happy to revisit this idea in a cleanup round, this patchset I feel like should focus on providing the basics of HDA at this point.

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -125,11 +276,14 @@ def map_phys_mem():
 
-    hugef = open(HUGEPAGE_FILE, "w+")
+    hugef_name = HUGEPAGE_FILE + str(stream_id)
+    hugef = open(hugef_name, "w+")
```
Probably not a big deal, but note that these pages are MUCH larger than we need, so it's a little wasteful to have one per stream.  Longer term it might be worth trying to optimize usage.  But then... a few dozen MB isn't that much on a modern system.

teburd replied:
I considered writing a quick'n'dirty allocator on top of a single huge page for just that reason... but thought there were better uses of time, for the moment.

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -125,11 +276,14 @@ def map_phys_mem():
+    log.info("type of mem is %s", str(type(mem)))
     global_mmaps.append(mem)
-    os.unlink(HUGEPAGE_FILE)
+    # helpful to sometimes comment out for debugging, can then cat the page
```
Not sure that comment is all that helpful.  I mean, yes, that's the way hugetlbfs works, the man page says it better than we can here.  If you want a mode to leave the pages behind for debugging, maybe we should add that to the interface?

teburd replied:
Might be worth while adding a flag to leave the huge pages, I found it helpful when debugging

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -329,6 +484,45 @@ def ipc_command(data, ext_data):
+        stream_id = ext_data & 0xFF
+        hda_str = hda_streams[stream_id]
+        hda_str.debug()
+        is_ramp_data = True
```
OK, you got me: I can't figure out what "ramp" means here. :)  Maybe just call this flag "valid" or something?

Also: maybe this is too much intelligence for the script?  If you just had a "read this offset from a stream buffer" command, you could do the validation in the C code where the buffer is generated, which is slower but IMHO a lot clearer and more flexible.

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -329,6 +484,45 @@ def ipc_command(data, ext_data):
+        ext_data = int(is_ramp_data)
+        log.info(f"Ext data to send back on ramp status {ext_data}")
+        send_msg = True
+    elif data == 10: # HDA HOST OUT SEND
```
Similarly, you could have a "write this word to this stream/offset" command and avoid the problem of having to synchronize the buffer formats between C and python.

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -343,7 +537,10 @@ def ipc_command(data, ext_data):
 async def main():
-    global hda, sd, dsp, hda_ostream_id
+    log.debug("Removing old hugetables")
+    runx("rm -rf /dev/hugepages/cavs_*")
```
What's this for?  The old scheme would detect and re-use the file if it was present.

teburd replied:
I wanted fresh and clean, every run. Which I didn't get without it. One of the many things I ended up falling into a rabbit hole on while debugging, probably removable now

andyross commented on soc/xtensa/intel_adsp/tools/cavstool.py:
```diff
@@ -343,7 +537,10 @@ def ipc_command(data, ext_data):
-    global hda, sd, dsp, hda_ostream_id
+    log.debug("Removing old hugetables")
+    runx("rm -rf /dev/hugepages/cavs_*")
+    #TODO this bit me, remove the globals, write a little FirmwareLoader class or something to contain.
```
Agreed.

teburd replied:
Todo in a cleanup round, when I move things to use HDAStream perhaps

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+#define TRANSFER_SIZE 256
+#define TRANSFER_COUNT 8
+
+__attribute__((section(".dma_buffers"), aligned(128))) uint8_t hda_fifo[FIFO_SIZE];
```
As mentioned earlier, everything will work here if you drop the section and just make an aligned static buffer.

teburd replied:
I'll try it, I had many issues, some of this stuff is left over because at the end of debugging I had a working version and wanted to move on to creating a dma driver on top of this. I tried reverting the linker script changes made above and the test failed so once it worked... I left well enough alone

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+	return true;
+}
+
+static void ipc_done(const struct device *dev, void *arg)
```
It should work to register a null handler, you don't have to have an empty function.

teburd replied:
I had some debug printks in here before, removable now. Noted for the cleanup round

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+	printk("dsp set buffer: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+	zassert_ok(res, "Expected set buffer to succeed");
+
+	k_sleep(K_MSEC(10));
```
What's this sleep for?

teburd replied:
Magic

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+
+		WAIT_FOR(cavs_ipc_send_message_sync(CAVS_HOST_DEV, IPCCMD_HDA_SEND,
+						    (STREAM_ID + 7) | (FIFO_SIZE << 8),
+						    IPC_TIMEOUT));
```
Not sure what's happening here?  This is going to iteratively retry the message send in a tight loop.  But it's a blocking call, it will wait on its own.   Why does it need a retry?  If you're getting spurious failures, there's likely some other bug.

Is it possible you send an async command earlier that hasn't completed yet?  If so, you want to poll on that and not the new message send.

teburd replied:
I tried removing these, and the tests fail presumably because a message isn't getting sent. So maybe another bug to hunt down then.

teburd replied:
Fixed

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+	bool is_ramp;
+
+	printk("smoke testing hda with fifo buffer at address %p, size %d\n",
+	       hda_fifo, FIFO_SIZE);
```
Maybe a little too much printk traffic here for my taste.  In test code no one's going to read it unless it fails, so it's generally best to put the logging into the zassert_*() cases (which have a printf format string and arguments at the end).

teburd replied:
For sure, was very helpful when I tested many, many times to see the screen blip by with my ramp data (0-255) or all zeros. Definitely overkill now that I have something working.

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+
+	res = cavs_hda_set_buffer(HDA_HOST_OUT_BASE, STREAM_ID, hda_fifo, FIFO_SIZE);
+	printk("dsp set buffer: "); cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+	zassert_ok(res, "Expected set buffer to succeed");
```
Shouldn't there be an assertion somewhere on the contents of the buffer, checking that the transfer moved the data correctly?

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+		cavs_hda_dbg("host_out", HDA_HOST_OUT_BASE, STREAM_ID);
+
+		/* The buffer is in the cached address range and must be invalidated
+		 * prior to reading.
```
Uh... no it isn't.  Or if it is that's a linkage bug.  Static data needs to be uncached always.  You get caching on the stack, and on symbols with an explicit __incoherent tag applied (**and** where you've taken care to make sure they're correctly aligned and padded and won't overlap a cache line with anyone else).  That's the core metaphor behind KERNEL_COHERENCE.  You can rely on everything except local variables being coherent.

teburd replied:
More leftover junk here, noted for the cleanup round

teburd replied:
Presumably I should be able to zassert_true(arch_mem_coherent(hda_fifo), "mem is coherent"); and that fails here for me

teburd replied:
It's because I had set MP_NUM_CPUS=1 in my test prj.conf trying to reduce what might be going wrong while debugging, this turns off the KERNEL_COHERENCE which is still probably wrong especially given the way HDA works.

teburd replied:
i've added tests to cover the NUM_CPUS=1 or the platform default to deal with this, as well as wrapping all the cache flush/invalidate bits in ifdefry

andyross commented on tests/boards/intel_adsp/hda/src/smoke.c:
```diff
@@ -0,0 +1,189 @@
+		z_xtensa_cache_inv(hda_fifo, FIFO_SIZE);
+		is_ramp = true;
+		for (int j = 0; j < FIFO_SIZE; j++) {
+			printk("hda_fifo[%d] = %d\n", j, hda_fifo[j]);
```
Ah, OK.  Here's the buffer test.  Ignore my comment above.

Just a `zassert_equal(hda_fifo[j], j, "blah blah blah...");` is sufficient here.  That will longjmp out of the test for you on failure.  No need to track a flag through the whole iteration.

teburd replied:
I want a long enough byte pattern, on both sides, that I can ensure things don't get corrupted. My initial test case actually sent a few bytes at a time in in a ring buffer with the data always containing the incremental byte values (0-255) in order, its pretty easy to spot if things have gone wrong, and no size limits by a predefined string then.

andyross reviewed (commented):
Super quick read through.  Really all this looks great to me.  Very clean.

Commit:
dma: Add read and write positions in dma_status

When using DMA with a circular buffer an application needs to know
where the hardware is in terms of its read and/or write positions.
Without this information, the circular buffer isn't very useful!

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
A CI failure is going on caused by the previous gpdma work and each soc enabling DMA_DW without actually using it.

Commit:
soc/intel_adsp: Require definitions for cavs_hda.h

Rather than defining them in the header, require a set of defines
be provided to cavs_hda.h as part of the expected input to the API.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Working on enabling this for all soc's, I have a PR that I can base that on and rework this a little use it, see #43343

teburd commented:
Now based on #43343 to avoid duplication of the device tree nodes and Kconfig defaults for intel_adsp family

Commit:
soc/intel_adsp: common testing code in cavstool.h

The IPC enum definition and WAIT_FOR macro are useful outside of
the board smoke tests for intel_adsp. They can be commonly used
by other board tests for a variety of peripherals that require
DSP and Host interaction (using cavstool).

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
soc/intel_adsp: Low level HDA driver and tests

Adds a header only low level driver for HDA streams along with smoke
tests to ensure basic host in and out stream functionality.

The tests require host side interaction. In cavstool a new HDAStream
class encapsulates somewhat a single stream and its registers. This
is manipulated in the tests using IPC with the Host ensuring that a
specific order of operations is done.

This low level driver allows testing certain hardware configurations
and flows with easy to use register dump debugging. It is not
intended to be the end API an application might use. That would be
a DMA driver using this.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
Rebased without changes

teburd commented:
Rebased with minor string changes (assert wording was slightly broken) and doc comments on cavs_hda_unused to address previous review comments.

ceolin commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,298 @@
+#define HDA_ALIGN_MASK 0xFFFFFF80
+
+
+#define HDA_ADDR(base, stream) (base + stream*HDA_REGBLOCK_SIZE)
```
`((base) + (stream)*HDA_REGBLOCK_SIZE)` ? It helps avoid wrong behavior if someone calls it `HDA_ADDR(base, stream + offset)`

ceolin reviewed (dismissed):
Awesome PR !

andyross reviewed (dismissed):
Refresh +1

mbolivar-nordic commented:
needs rebase?

Commit:
dma/cavs_hda: Support source/dest width setting

Support adjusting the sample size to 16bits if the source
or destination width is set 3 or less.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
dma/cavs_hda: Provide pending/free byte lengths

In status the pending/free byte lengths are now provided

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

abonislawski commented on dts/xtensa/intel/intel_cavs.dtsi:
```diff
@@ -31,5 +31,25 @@
 			status = "okay";
 		};
+
+		hda_host_out: dma@0x72800 {
```
Probably just @72800?
Warning (simple_bus_reg): /soc/dma@0x72800: simple-bus unit address format error, expected "72800"

abonislawski commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+#define DGBSP(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x18))
+
+/* Gateway Minimum Buffer Size */
+#define DGMBS(base, stream) ((volatile uint32_t *)(HDA_ADDR(base, stream) + 0x1c))
```
I was checking registry values in real audio playback scenario and see no use of DGMBS, this one is the only difference (SOF will put buffer size here).
Not a big deal, probably doesnt matter, just FYI

abonislawski reviewed (commented):
I can confirm that it works great with SOF audio

Commit:
soc/cavs: Fix warning in intel_cavs.dtsi

Incorrectly formatted the address with the 0x hex prefix. Fixed

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

marc-hb commented on soc/xtensa/intel_adsp/common/include/cavs_hda.h:
```diff
@@ -0,0 +1,303 @@
+#define DGCS_SCS BIT(31) /* Sample container size */
+#define DGCS_GEN BIT(26) /* Gateway Enable */
+#define DGCS_L1ETP BIT(25) /* L1 Enter Prevent */
+#define DGCS_L1EXP BIT(25) /* L1 Exit Prevent */
```
Fixed in
- #58922
